### PR TITLE
[Placement Group] Support RayPlacementGroupError #10508

### LINKED ISSUE #10508: [Placement Group] Support RayPlacementGroupError
### Describe your feature request
When tasks are failed due to placement group removal, we should raise `RayPlacementGroupError` instead of worker dead error or actor dead error.

### PR DESCRIPTION
**Pull request by oliverhu**
## Why are these changes needed?
Fix https://github.com/ray-project/ray/issues/10508

Also:
- Refactored `ProcessDisconnectClientMessage` to an msg processing function and an util function. A lot of message passing is from the same file.
- Removed `intentional_disconnect_client`, replaced with a `ClientDisconnectType` to provide more information on disconnect reasoning.
- Renamed `DestroyWorker` to `DisconnectAndKillWorker` to bring more clarity.

## Related issue number
https://github.com/ray-project/ray/issues/10508

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by oliverhu** [comment]
Reverted changes in https://github.com/ray-project/ray/pull/12821/files that deleted `intention_disconnect` message type, and `message_handler` signature changes, as those changes are fairly risky and not related to this specific issue. Will open another discussion thread on that.

**Comment by oliverhu** [comment]
```> pytest ./ray/tests/test_failure.py
Test session starts (platform: darwin, Python 3.8.3, pytest 5.4.3, pytest-sugar 0.9.4)
rootdir: /Users/khu/ray/ray/python
plugins: sugar-0.9.4, rerunfailures-9.1, asyncio-0.14.0, timeout-1.4.2
collecting ...
 ray/tests/test_failure.py ✓✓✓✓✓✓✓✓✓✓✓✓✓✓✓✓sss✓✓✓✓✓✓✓✓✓✓✓✓✓✓✓✓✓✓✓ 84% ████████▌
                           ✓✓✓✓✓                                                                          ✓✓✓✓✓✓                                                        ✓✓✓✓✓✓✓                                                                                            100% ██████████
============================================================= warnings summary ==============================================================
ray/tests/test_failure.py::test_failed_task
  /Users/khu/opt/anaconda3/lib/python3.8/site-packages/aiohttp/helpers.py:107: DeprecationWarning: "@coroutine" decorator is deprecated since Python 3.8, use "async def" instead
    def noop(*args, **kwargs):  # type: ignore

ray/tests/test_failure.py: 23 tests with warnings
  /Users/khu/ray/ray/python/ray/exceptions.py:87: DeprecationWarning: PY_SSIZE_T_CLEAN will be required for '#' formats
    self.proctitle = setproctitle.getproctitle()

-- Docs: https://docs.pytest.org/en/latest/warnings.html

Results (247.39s):
      42 passed
       3 skipped
```

File: src/ray/protobuf/common.proto [hunk 1]
```diff
@@ -133,6 +133,8 @@ enum ErrorType {
   TASK_CANCELLED = 5;
   // Inidicates that creating the GCS service failed to create the actor.
   ACTOR_CREATION_FAILED = 6;
+  // Indicates failure due to placement group
```
Comment (ffbin) [inline src/ray/protobuf/common.proto:136]:
```suggestion
  // Indicates failure due to placement group.
```

File: src/ray/protobuf/common.proto [hunk 2]
```diff
@@ -442,3 +444,15 @@ message MetricPoint {
...
+// Type of a worker exit.
+enum ClientDisconnectType {
+  // Unintended worker exit.
+  UNEXPECTED_EXIT = 0;
+  // Intended worker exit.
+  FINISHED = 1;
+  // Worker exit due to resource bundle release.
+  UNUSED_RESOURCE_RELEASED = 2;
+  // Worker exit due to placement group removal.
+  PLACEGROUP_REMOVED = 3;
```
Comment (ffbin) [inline src/ray/protobuf/common.proto:457]:
```suggestion
  PLACEMENT_GROUP_REMOVED = 3;
```

File: src/ray/raylet/node_manager.cc [hunk 3]
```diff
@@ -347,11 +347,15 @@ void NodeManager::KillWorker(std::shared_ptr<WorkerInterface> worker) {
...
+void NodeManager::DisconnectAndKillWorker(std::shared_ptr<WorkerInterface> worker,
+                                          rpc::ClientDisconnectType disconnect_type) {
+  // Used to destroy a worker when its bundle resource is released (unused or
+  // placementgroup is deleted.)
   // We should disconnect the client first. Otherwise, we'll remove bundle resources
   // before actual resources are returned. Subsequent disconnect request that comes
   // due to worker dead will be ignored.
-  ProcessDisconnectClientMessage(worker->Connection(), /* intentional exit */ true);
+  DisconnectClient(worker->Connection(), disconnect_type);
+  RAY_LOG(ERROR) << "DisconnectAndKillWorker" << disconnect_type;
```
Comment (ffbin) [inline src/ray/raylet/node_manager.cc:358]: This log is just for debug?

File: src/ray/raylet/node_manager.cc [hunk 4]
```diff
@@ -1354,8 +1358,9 @@ void NodeManager::HandleWorkerAvailable(const std::shared_ptr<WorkerInterface> &
   }
 }
 
-void NodeManager::ProcessDisconnectClientMessage(
-    const std::shared_ptr<ClientConnection> &client, bool intentional_disconnect) {
+void NodeManager::DisconnectClient(const std::shared_ptr<ClientConnection> &client,
+                                   rpc::ClientDisconnectType disconnect_type) {
+  RAY_LOG(ERROR) << "DisconnectClient*****" << disconnect_type;
```
Comment (ffbin) [inline src/ray/raylet/node_manager.cc:1363]: Remove debug log.

File: src/ray/raylet/node_manager.cc [hunk 5]
```diff
@@ -1400,32 +1404,48 @@ void NodeManager::ProcessDisconnectClientMessage(
...
+      switch (disconnect_type) {
+      case rpc::ClientDisconnectType::UNEXPECTED_EXIT: {
+        type_str = "worker_died";
+        error_message << "A worker died or was killed while executing task " << task_id
+                      << ".";
+      } break;
+      case rpc::ClientDisconnectType::PLACEGROUP_REMOVED: {
+        type_str = "placement_group";
+        error_message << "A worker was killed while executing task " << task_id
+                      << " due to placement group removal"
```
Comment (ffbin) [inline src/ray/raylet/node_manager.cc:1432]:
```suggestion
                      << " due to placement group removal."
```

File: src/ray/raylet/node_manager.cc [hunk 6]
```diff
@@ -1400,32 +1404,48 @@ void NodeManager::ProcessDisconnectClientMessage(
...
     if ((!task_id.IsNil() || !actor_id.IsNil()) && !worker->IsDead()) {
+      // If the worker was running a task or actor, clean up the task and push an
+      // error to the driver, unless the worker is already dead.
+      // TODO(rkn): Define this constant somewhere else.
+      std::string type_str;
+      std::ostringstream error_message;
+
+      bool intentional_exit = false;
+      switch (disconnect_type) {
+      case rpc::ClientDisconnectType::UNEXPECTED_EXIT: {
```
Comment (ffbin) [inline src/ray/raylet/node_manager.cc:1424]: {} is not needed.

File: src/ray/raylet/node_manager.cc [hunk 7]
```diff
@@ -1838,7 +1868,7 @@ void NodeManager::HandleReturnWorker(const rpc::ReturnWorkerRequest &request,
 
   if (worker) {
     if (request.disconnect_worker()) {
-      ProcessDisconnectClientMessage(worker->Connection());
+      DisconnectClient(worker->Connection(), rpc::ClientDisconnectType::UNEXPECTED_EXIT);
```
Comment (ffbin) [inline src/ray/raylet/node_manager.cc:1871]:
```suggestion
      DisconnectClient(worker->Connection());
```

File: src/ray/raylet/node_manager.h [hunk 8]
```diff
@@ -472,7 +483,7 @@ class NodeManager : public rpc::NodeManagerServiceHandler {
   /// client.
   ///
   /// \param client The client that sent the message.
-  /// \param intentional_disconnect Whether the client was intentionally disconnected.
+  /// \param message_data A pointer to the message data.
```
Comment (ffbin) [inline src/ray/raylet/node_manager.h:486]: Comments and code are inconsistent.

File: src/ray/raylet/node_manager.cc (hunk 4 above)
Comment (oliverhu) [reply src/ray/raylet/node_manager.cc:1363]: this commit was not pushed to the branch 🤦  fixed

File: src/ray/raylet/node_manager.cc [hunk 9]
```diff
@@ -1255,7 +1258,7 @@ void NodeManager::ProcessRegisterClientRequestMessage(
         static_cast<int64_t>(protocol::MessageType::RegisterClientReply), fbb.GetSize(),
         fbb.GetBufferPointer(), [this, client](const ray::Status &status) {
           if (!status.ok()) {
-            ProcessDisconnectClientMessage(client);
+            DisconnectClient(client, rpc::ClientDisconnectType::UNEXPECTED_EXIT);
```
Comment (ffbin) [inline src/ray/raylet/node_manager.cc:1261]:
```suggestion
            DisconnectClient(client);
```

File: src/ray/raylet/node_manager.cc [hunk 10]
```diff
@@ -1400,32 +1402,48 @@ void NodeManager::ProcessDisconnectClientMessage(
...
+        error_message << "A worker died or was killed while executing task " << task_id
+                      << ".";
+        break;
+      case rpc::ClientDisconnectType::PLACEMENT_GROUP_REMOVED:
+        type_str = "placement_group";
+        error_message << "A worker was killed while executing task " << task_id
+                      << " due to placement group removal."
+                      << ".";
+        break;
+      default:
```
Comment (ffbin) [inline src/ray/raylet/node_manager.cc:1433]: case rpc::ClientDisconnectType::FINISHED:
case rpc::ClientDisconnectType:: UNUSED_RESOURCE_RELEASED:
  intentional_exit = true;

Comment (ffbin) [reply src/ray/raylet/node_manager.cc:1433]: Or you can add a comment to explain what default is.

File: src/ray/raylet/node_manager.h [hunk 11]
```diff
@@ -384,12 +384,23 @@ class NodeManager : public rpc::NodeManagerServiceHandler {
   /// \return Void.
   void KillWorker(std::shared_ptr<WorkerInterface> worker);
 
-  /// Destroy a worker.
+  /// Disconnect and kill a worker.
   /// We will disconnect the worker connection first and then kill the worker.
   ///
-  /// \param worker The worker to destroy.
+  /// \param worker The worker to stop.
```
Comment (ffbin) [inline src/ray/raylet/node_manager.h:390]: Add `/// \param disconnect_type ....`

File: src/ray/raylet/node_manager.cc [hunk 12]
```diff
@@ -347,11 +347,14 @@ void NodeManager::KillWorker(std::shared_ptr<WorkerInterface> worker) {
   });
 }
 
-void NodeManager::DestroyWorker(std::shared_ptr<WorkerInterface> worker) {
+void NodeManager::DisconnectAndKillWorker(std::shared_ptr<WorkerInterface> worker,
+                                          rpc::ClientDisconnectType disconnect_type) {
+  // Used to destroy a worker when its bundle resource is released (unused or
+  // placementgroup is deleted.)
```
Comment (ffbin) [inline src/ray/raylet/node_manager.cc:353]:
```suggestion
  // placement group is deleted).
```

File: src/ray/raylet/node_manager.cc [hunk 13]
```diff
@@ -1400,32 +1402,52 @@ void NodeManager::ProcessDisconnectClientMessage(
...
+      case rpc::ClientDisconnectType::UNEXPECTED_EXIT:
+        type_str = "worker_died";
+        error_message << "A worker died or was killed while executing task " << task_id
+                      << ".";
+        break;
+      case rpc::ClientDisconnectType::PLACEMENT_GROUP_REMOVED:
+        type_str = "placement_group";
+        error_message << "A worker was killed while executing task " << task_id
+                      << " due to placement group removal."
+                      << ".";
```
Comment (ffbin) [inline src/ray/raylet/node_manager.cc:1431]: `<< "."` is redundant

**Review by ffbin — APPROVED** [review]
LGTM, just a small comment.

**Review by rkooo567 — APPROVED** [review]
Can you add tests to `test_placement_group`?

**Comment by oliverhu** [comment]
@rkooo567 and I discussed further on this PR, current changes are not enough to cover all cases (direct task and actor task). and we might have to subscribe to worker changes, and make changes to ActorTableData to track actor exit status. We will park this PR for a bit of time..

**Comment by bveeramani** [comment]
## :bangbang: ACTION REQUIRED :bangbang:

We've switched our code formatter from YAPF to Black (see #21311).

To prevent issues with merging your code, here's what you'll need to do:

1. Install Black
```
pip install -I black==21.12b0
```

2. Format changed files with Black
```
curl -o format-changed.sh https://gist.githubusercontent.com/bveeramani/42ef0e9e387b755a8a735b084af976f2/raw/7631276790765d555c423b8db2b679fd957b984a/format-changed.sh
chmod +x ./format-changed.sh
./format-changed.sh
rm format-changed.sh
```

3. Commit your changes.
```
git add --all
git commit -m "Format Python code with Black"
```

4. Merge master into your branch.
```
git pull upstream master
```

5. Resolve merge conflicts (if necessary).

After running these steps, you'll have the updated `format.sh`.

**Pull request closed without merging**